### ISSUE TITLE
[Phase 2] Missing Value Imputation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/real-estate-valuation-engine#4]
## Summary

Implement iterative imputation for missing values using feature-specific models. Property data commonly has missing values that are correlated (e.g., missing sqft can be estimated from bedrooms).

**Part of:** #1
**Depends on:** #3 (Phase 1)

## Objectives

- [ ] Implement `MissingValueHandler` class
- [ ] Use IterativeImputer with BayesianRidge for numeric features
- [ ] Use SimpleImputer (mode) for categorical features
- [ ] Track imputation statistics
- [ ] Ensure no data leakage (fit on train, transform on test)
- [ ] Write unit tests

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/data/imputer.py` | 1-100 | Implement | MissingValueHandler class |
| `tests/data/test_imputer.py` | NEW | Create | Unit tests |

### MissingValueHandler Class

```python
# src/data/imputer.py
import numpy as np
import pandas as pd
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer, SimpleImputer
from sklearn.linear_model import BayesianRidge
from typing import List

class MissingValueHandler:
    """Handle missing values with feature-appropriate strategies."""

    def __init__(
        self,
        numeric_features: List[str],
        categorical_features: List[str],
        max_iter: int = 10,
        random_state: int = 42
    ):
        self.numeric_features = numeric_features
        self.categorical_features = categorical_features
        self.max_iter = max_iter
        self.random_state = random_state

        self._numeric_imputer: IterativeImputer = None
        self._categorical_imputer: SimpleImputer = None
        self.imputation_stats_: dict = {}

    def fit(self, df: pd.DataFrame) -> 'MissingValueHandler':
        """Fit imputers on training data."""
        # Record missing percentages before
        self.imputation_stats_['missing_before'] = df.isnull().mean().to_dict()

        # Fit numeric imputer
        if self.numeric_features:
            self._numeric_imputer = IterativeImputer(
                estimator=BayesianRidge(),
                max_iter=self.max_iter,
                random_state=self.random_state
            )
            self._numeric_imputer.fit(df[self.numeric_features])

        # Fit categorical imputer
        if self.categorical_features:
            self._categorical_imputer = SimpleImputer(strategy='most_frequent')
            self._categorical_imputer.fit(df[self.categorical_features])

        return self

    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """Apply imputation to data."""
        df = df.copy()

        if self.numeric_features and self._numeric_imputer:
            df[self.numeric_features] = self._numeric_imputer.transform(
                df[self.numeric_features]
            )

        if self.categorical_features and self._categorical_imputer:
            df[self.categorical_features] = self._categorical_imputer.transform(
                df[self.categorical_features]
            )

        self.imputation_stats_['missing_after'] = df.isnull().mean().to_dict()
        return df

    def fit_transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """Fit and transform in one step."""
        return self.fit(df).transform(df)
```

### Test: No Data Leakage

```python
# tests/data/test_imputer.py
import pytest
import pandas as pd
import numpy as np
from src.data.imputer import MissingValueHandler

def test_imputer_no_leakage():
    """Imputer should only use training data statistics."""
    # Train data with low values
    train_df = pd.DataFrame({
        'sqft': [1000, None, 1100, 1050, None],
        'price': [100, 110, 105, 108, 102]
    })

    # Test data with high values (should NOT influence imputation)
    test_df = pd.DataFrame({
        'sqft': [None, 5000],  # Missing should be ~1000, not 5000
        'price': [500, 510]
    })

    handler = MissingValueHandler(
        numeric_features=['sqft', 'price'],
        categorical_features=[]
    )

    # Fit on train only
    handler.fit(train_df)

    # Transform test
    test_imputed = handler.transform(test_df)

    # Imputed sqft should be close to train mean (~1050), not test (5000)
    assert test_imputed['sqft'].iloc[0] < 2000
```

## Verification

```bash
pytest tests/data/test_imputer.py -v

python -c "
from src.data.imputer import MissingValueHandler
import pandas as pd

df = pd.DataFrame({
    'sqft': [1000, None, 1200],
    'bedrooms': [3, 2, None],
    'property_type': ['house', None, 'condo']
})

handler = MissingValueHandler(
    numeric_features=['sqft', 'bedrooms'],
    categorical_features=['property_type']
)

imputed = handler.fit_transform(df)
print(imputed)
print(f'Missing after: {imputed.isnull().sum().sum()}')
"
```

## Technical Challenges

1. **IterativeImputer convergence** - May not converge with high missing rates
   - Solution: Set `max_iter=10`, check convergence warnings
2. **Data leakage** - Must fit only on training data
   - Solution: Separate fit() and transform() methods

## Definition of Done

- [ ] Zero missing values after imputation
- [ ] Train/test split respected (no leakage)
- [ ] Imputation reproducible with random_state
- [ ] Imputation statistics tracked
- [ ] All unit tests pass

## Labels

`phase-2`, `data`, `enhancement`
